### PR TITLE
Issue #728 add Neo4j provider

### PR DESCRIPTION
**Pull request by thimmwork**
I created a neo4j provider.
Tests were mostly copied from jdbc modules and merged into a single integration test class.

Neo4j uses read-committed isolation, which may lead to lost updates when not taken care of.
I used explicit node locking as described in [Neo4j's documentation about lost updates](https://neo4j.com/docs/java-reference/current/transaction-management/#transactions-isolation-lostupdates) to accommodate for this. FuzzTester tells me this is fine.

Let me know what you think :)

### CODE REVIEW

**Comment by lukas-krecan** [comment]
Hi, thanks a lot, I will take a look at it in next few days.

File: pom.xml [hunk 1]
```diff
@@ -44,6 +44,7 @@
         <module>providers/jdbc/shedlock-provider-jdbc-micronaut</module>
         <module>providers/mongo/shedlock-provider-mongo</module>
         <module>providers/mongo/shedlock-provider-mongo-reactivestreams</module>
+        <module>providers/neo4j</module>
```
Comment (lukas-krecan) [inline pom.xml:47]: I would move this to a submodule. I assume there are multiple libraries you can use to access neo4j so we should make space for potential different implementation.

File: providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.neo4j.driver.Session;
+import org.neo4j.driver.Transaction;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Objects.requireNonNull;
+
+public class Neo4jStorageAccessor extends AbstractStorageAccessor {
```
Comment (lukas-krecan) [inline providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java:34]: Please make it package-protected.

File: providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    public Neo4jStorageAccessor(@NonNull Driver driver, @NonNull String collectionName) {
+        this.collectionName = requireNonNull(collectionName, "collectioneName can not be null");
+        this.driver = requireNonNull(driver, "driver can not be null");
+        createLockNameUniqueConstraint(driver);
+    }
+
+    private void createLockNameUniqueConstraint(Driver driver) {
+        try (Session session = driver.session();
+             Transaction transaction = session.beginTransaction()) {
+            transaction.run("CREATE CONSTRAINT UNIQUE_shedlock_name IF NOT EXISTS ON (lock:shedlock) ASSERT lock.name IS UNIQUE");
```
Comment (lukas-krecan) [inline providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java:47]: Should use `collectionName`

File: providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    @NonNull
+    protected Neo4jStorageAccessor getAccessor(boolean usingDbTime) {
+        return new Neo4jStorageAccessor(this.testUtils.getDriver(), "shedlock");
+    }
+
+
+    @Test
+    @Disabled
+    public void shouldNotFailIfKeyNameTooLong() {
+        LockConfiguration configuration = lockConfig("lock name that is too long Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.");
```
Comment (lukas-krecan) [inline providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java:226]: Why is this disabled?

File: providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    @Override
+    public boolean updateRecord(@NonNull LockConfiguration lockConfiguration) {
+        String cypher = String.format("MATCH (lock:%s) " +
+                "WHERE lock.name = $lockName AND lock.lock_until <= $now " +
+                "SET lock._LOCK_ = true " +
+                "WITH lock as l " +
+                "WHERE l.lock_until <= $now " +
+                "SET l.lock_until = $lockUntil, l.locked_at = $now, l.locked_by = $lockedBy " +
+                "REMOVE l._LOCK_ ", collectionName);
+        Map<String, Object> parameters = new HashMap<>();
```
Comment (lukas-krecan) [inline providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java:76]: Duplicated code

File: providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    public void unlock(@NonNull LockConfiguration lockConfiguration) {
+        String cypher = String.format("MATCH (lock:%s) WHERE lock.name = $lockName " +
+            "SET lock.lock_until = $lockUntil ", collectionName);
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("lockName", lockConfiguration.getName());
+        parameters.put("lockUntil", lockConfiguration.getUnlockTime().toString());
+        executeCommand(cypher, statement -> null, parameters, this::handleUnlockException);
+    }
+
+    protected <T> T executeCommand(
```
Comment (lukas-krecan) [inline providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java:120]: Why protected?

File: providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        container.start();
+    }
+
+    @AfterAll
+    public void shutDownDb() {
+        container.stop();
+    }
+
+    protected boolean useDbTime() {
+        return false;
```
Comment (lukas-krecan) [inline providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java:82]: Why do we have it here?

**Review by lukas-krecan — COMMENTED** [review]
Good job, I just have few comments/questions.

File: providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java (hunk 4 above)
Comment (thimmwork) [reply providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java:226]: Initially it was disabled in JDBC module which I used as copy source. But a test like this does not make sense as the limit is machine dependent and equal to the limit of Java String. I removed it.

File: providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java (hunk 6 above)
Comment (thimmwork) [reply providers/neo4j/src/main/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jStorageAccessor.java:120]: changed to private

File: providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java (hunk 7 above)
Comment (thimmwork) [reply providers/neo4j/src/test/java/net/javacrumbs/shedlock/provider/neo4j/Neo4jLockProviderIntegrationTest.java:82]: copy-pasta. I removed it.

File: pom.xml (hunk 1 above)
Comment (thimmwork) [reply pom.xml:47]: I moved it to a subfolder with no additional intermediate maven submodule - now it's structured like the other folders.

**Comment by lukas-krecan** [comment]
Thanks, looks good

**Pull request merged**